tests/reminder_test.php

```diff
@@ -0,0 +1,139 @@
+        $this->getDataGenerator()->enrol_user($this->users[2]->id, $this->course->id, $teacherrole->id);
+    }
+
+    /**
+     * Test for sending the reminder after duedate.
+     *
+     * @return void
+     */
+    public function test_reminder_duedate() {
+        global $DB;
+
+        $this->resetAfterTest();
+
+        $boardmanager = new boardmanager($this->kanban->cmid);
+        $boardid = $boardmanager->create_board();
+        $boardmanager->load_board($boardid);
+        $columnids = explode(',', $boardmanager->get_board()->sequence);
+        $boardmanager->add_card($columnids[0], 0, ['title' => 'Testcard', 'duedate' => time() - 10000]);
+
+        $reminder = new \mod_kanban\task\reminder();
+        $reminder->execute();
+
+        $noreply = \core_user::get_noreply_user();
+
+        // No one is assigned to the card, so there should be no notifications.
+
+        for ($i = 0; $i < 3; $i++) {
+            $this->assertEquals(
+                0,
+                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
+            );
+        }
+
+        $cardid = $boardmanager->add_card($columnids[0], 0, ['title' => 'Testcard 2', 'duedate' => time() - 10000]);
+        $boardmanager->assign_user($cardid, $this->users[0]->id);
+
+        $notificationcount = [];
+
+        for ($i = 0; $i < 3; $i++) {
+            $notificationcount[$i] = $DB->count_records(
+                'notifications',
+                ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id]
+            );
+        }
+
+        $reminder->execute();
+
+        // User 0 should have one additional notification.
+
+        for ($i = 0; $i < 3; $i++) {
+            $this->assertEquals(
+                $notificationcount[$i] + (int)($i == 0),
+                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
+            );
+        }
+
+        $cardid = $boardmanager->add_card(
+            $columnids[0],
+            0,
+            ['title' => 'Testcard 3', 'duedate' => time() + 100000000, 'reminderdate' => time() - 1000]
+        );
+        $boardmanager->assign_user($cardid, $this->users[0]->id);
+        $boardmanager->assign_user($cardid, $this->users[2]->id);
+
+        for ($i = 0; $i < 3; $i++) {
+            $notificationcount[$i] = $DB->count_records(
+                'notifications',
+                ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id]
+            );
+        }
+
+        $reminder->execute();
+
+        // User 0 and 2 should have one additional notification.
+
+        for ($i = 0; $i < 3; $i++) {
+            $this->assertEquals(
+                $notificationcount[$i] + (int)($i % 2 == 0),
+                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
+            );
+        }
```

Comment:
I like line 134 very much, but it's really horrible to read :-D

```suggestion
        for ($i = 0; $i < 3; $i++) {
            // Zero notifications expected for user 1, one notification each expected for the other two users.
            $expectednotificationcount = $i === 1 ? 0 : 1;
            $this->assertEquals(
                $expectednotificationcount,
                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
            );
        }
```

Reply:
Your solution is fine as well :+1: